File: Game.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package game;

import java.util.Scanner;

/**
 *
 * @author Nina
 */
public class Game {
    
    static int index = -1;
    static int count = 0;
    
    public static boolean nextBoolean() {
        Scanner s = new Scanner(System.in);
        return s.nextBoolean();
    }
    
    public static int nextInteger() {
        Scanner s = new Scanner(System.in);
        return s.nextInt();
    }
    
    public static Guess nextTurn(Player[] p, Guess guess) {
        if (guess.count()==1 && guess.value()==0) {
            return newGame(p, guess);
        } else {
            if (index<p.length-1) {
                index++;
            } else {
                index = 0;
            }
            if (p[index].isActive()) {
                return p[index].play(p, guess, index);
            } else {
                return guess;
            }
        }
    }
    
    public static void winnerPrinter(int winner, Player[] p) {
        System.out.println("Player " + (winner)); 
        p[winner-1].win();
    }
    
    public static void newRoundPrinter(Player[] p) {
        System.out.println("New round");     
        System.out.println("Player " + (index+1));
        System.out.println("Your dices:");
        for (int d : p[index].getDices()) {
            System.out.print(d + " ");
        }
        System.out.println();
    }
    
    public static Guess newGame(Player[] p, Guess guess) { //pozmenit
        int playing = 0;
        int winner = -1;
        for (int i=0; i<p.length; i++) {
            if (p[i].isActive()) {
                playing++;
                winner = i;
            }            
        }
        
        if (index==-1) {
            int i = 1;
            for (Player pl : p) {
                pl.wannaRoll(i);
                i++;
            }
        }
        
        if (playing<2) {
            winnerPrinter(winner+1, p);
            return new Guess(0,0);
        } else {
            if (index<p.length-1) {
                index++;
            } else {
                index = 0;
            }
            newRoundPrinter(p);
            p[index].askShow();
            return p[index].raise(guess);
        }
    }
    
    public static Player[] createGame(int n, int m) {
        Player[] players = new Player[n];        
        for (int i=0; i<n; i++) {
            players[i] = new Player(m);
            count += m;
        }
        return players;
    }
    
    public static void game(Player[] players, Guess guess) {
        guess = newGame(players, guess);
        while(true) {
            guess = nextTurn(players, guess);
            if (guess.count()==0) {
                break;
            }
        }        
    }
    

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        // TODO code application logic here
        int n, m;
        
        System.out.println("Enter the number of players:");
        while(true) {
            n = nextInteger();
            if (n<2) {
                System.out.println("At least 2 players are required. Please try again.");
            } else {
                break;
            }
        }
        
        System.out.println("Enter the number of dices:");
        while(true) {
            m = nextInteger();
            if (m<1) {
                System.out.println("You need at least 1 dice to play. Please try again.");
            } else {
                break;
            }
        }
        
        System.out.println("Game starts");
        game(createGame(n, m), new Guess(1,0));
        System.out.println("Game ends");
    }
    
}

File: game/public/Game.js
let Player = null;
let Guess = null;
if (typeof module !== "undefined") {
    Guess = require("./Guess");
    Player = require("./Player");
}

class Game {

    static nextBoolean(bool, socket, next, args, gameID) {
        var f = document.createElement("div");
        var b1 = document.createElement("button");
        b1.innerHTML = "Yes";
        b1.style = "height: 30px; width: 60px; margin: 10px;";
        var b2 = document.createElement("button");
        b2.innerHTML = "No";
        b2.style = "height: 30px; width: 60px; margin: 10px;";
        f.appendChild(b1);
        f.appendChild(b2);
        b1.onclick = function () {
            f.parentElement.removeChild(f);
            bool.value = true;
            document.getElementById("paragraph").innerHTML = "";
            socket.emit('createMessage', {
                code: "answer",
                value: { answer: bool.value, continue: next, args: args, gameID: gameID }
            });
        };

        b2.onclick = function () {
            f.parentElement.removeChild(f);
            bool.value = false;
            document.getElementById("paragraph").innerHTML = "";
            socket.emit('createMessage', {
                code: "answer",
                value: { answer: bool.value, continue: next, args: args, gameID: gameID }
            });
        };
        document.getElementById("paragraph").appendChild(f);
    }

    static nextInteger(num, bigger, smaller, message, next, args) {
        var f = document.createElement("div");
        var b = document.createElement("button");
        b.type = "submit";
        b.style = "height: 20px; width: 60px; margin: 5px;";
        b.innerHTML = "Submit";
        var s = document.createElement("input");
        f.appendChild(s);
        f.appendChild(b);
        var x;
        b.onclick = function () {
            x = s.value;
            if (!Number.isInteger(Number(x))) {
                alert("Must input integer.");
                f.parentElement.removeChild(f);
                Game.nextInteger(num, bigger, smaller, message, next, args);
            } else {
                f.parentElement.removeChild(f);
                num.value = Number(x);
                Game.check(num, bigger, smaller, message, next, args);
            }
        };
        s.addEventListener("keypress", function(event) {
            if (event.keyCode == 13) {
                b.click();
            }
        });
        document.getElementById("paragraph").appendChild(f);
    }

    static nextTurn(game) {
        if (game.guess.count == 1 && game.guess.value == 0 || game.ended) {
            game.newTurn = true;
            Game.newGame(game);
        } else {
            if (game.index < game.players.length-1) {
                game.index++;
            } else {
                game.index = 0;
            }
            if (game.players[game.index].isActive()) {
                game.players[game.index].play(game.players, game.guess);
            } else {
                Game.nextTurn(game);
            }
        }
    }

    static winnerPrinter(winner, p) {
        p[winner].win();
    }

    static newRoundPrinter(game) {
        for (var i=0; i<game.players.length; i++) {
            if (game.players[i].isActive()) {
                game.clients[i].emit('newMessage', {
                    code: 2,
                    value: "New round"
                });
            }
        }
        while(!game.players[game.index].isActive()) {
            if (game.index < game.players.length-1) {
                game.index++;
            } else {
                game.index = 0;
            }
        }
        game.players[game.index].askShow();
    }

    static newGame(game) {
        var p = game.players;
        var playing = 0;
        var winner = -1;
        for (var i = 0; i < p.length; i++) {
            if (p[i].isActive()) {
                playing++;
                winner = i;
            }
        }
        if (playing < 2) {
            game.ended = true;
            game.clients[winner].emit('newMessage', {
                code: 5,
                value: { winner: p[winner].dices }
            });
            Game.winnerPrinter(winner, p);
        } else {
            for (var pl of p) {
                if (pl.isActive()) {
                    pl.wannaRoll();
                }
            }
        }
    }

    static newGame2(game) {
        if (game.index < game.players.length-1) {
            game.index++;
        } else {
            game.index = 0;
        }
        Game.newRoundPrinter(game);
    }

    static createGame(n, m) {
        var players = new Array();
        for (var i = 0; i < n; i++) {
            players[i] = new Player(m);
        }
        return players;
    }

    static game(game) {
        Game.newGame(game);
    }

    static check(num, bigger, smaller, message, next, args) {
        if (num.value <= bigger || num.value > smaller) {
            alert(message);
            Game.nextInteger(num, bigger, smaller, message, next, args);
        } else {
            document.getElementById("paragraph").innerHTML = "";
            next(args);
        }
    }

    static send(args) {
        console.log(args);
        args.socket.emit('createMessage', {
            code: "do",
            value: { function: args.function, num: args.value.value, args: args.args, gameID: args.gameID }
        });
    }

    static gameStarter(args) {
        args.socket.emit('createMessage', {
            code: "start",
            value: { n: args.n1.value, m: args.n2.value, name: sessionStorage.getItem("name") }
        });
        if (args.n1.value == 2) {
            document.getElementById("paragraph").innerHTML = "Waiting for " + (args.n1.value-1) + " player to connect.";
        } else if (args.n1.value > 2) {
            document.getElementById("paragraph").innerHTML = "Waiting for " + (args.n1.value-1) + " players to connect.";
        } else {
            document.getElementById("paragraph").innerHTML = "Game";
        }
        alert("Game starts");
    }

    static continue(args) {
        var m = new Number();
        var s = "You need at least 1 and at most 50 dices to play. Please try again.";
        document.getElementById("paragraph").innerHTML = "Enter the number of dices:";
        Game.nextInteger(m, 0, 50, s, Game.gameStarter, { n1: args.n, n2: m, socket: args.socket });
    }

    static main(socket) {
        // TODO code application logic here
        var args = { n: new Number(), socket: socket }
        var s = "At least 2 and at most 10 players are required. Please try again.";
        document.getElementById("paragraph").innerHTML = "Enter the number of players:";
        Game.nextInteger(args.n, 1, 10, s, Game.continue, args);
    }

}


if (typeof module !== "undefined" && module.exports) {
    module.exports = Game;
}
